### DISCUSSION TITLE
First Lap Time

### DISCUSSION THREAD

**Discussion by IceDragons37** [theOehrly/Fast-F1#189] [Q&A]
import pandas as pd
import fastf1 as ff1
from fastf1 import plotting

race = ff1.get_session(Year-1, Grand_Prix, 'R')
race.load()
laps = race.laps

ff1.Cache.enable_cache('cache')
pd.options.mode.chained_assignment = None

laps_rus = laps.pick_driver('RUS')  # pick a driver
lap = laps_rus[laps_rus['LapNumber'] == 1].iloc[0]
print(lap)

When I print the info, the lap time and the Sector1 time are stated as "NaT". Is there a way to find the lap time and the time taken to complete the 1st sector?

**Comment by theOehrly** [comment]
No, there isn't because the sector 1 time and lap time do not exist for the first lap of a race.

You can take a look at https://www.fia.com/sites/default/files/decision-document/2022%20Spanish%20Grand%20Prix%20-%20Event%20Notes%20-%20Circuit%20Map%20v2.pdf for example.
The control line marks the beginning of the lap (i.e. the beginning of sector 1) but the cars are lined up at the start line. So they are already inside sector 1. They're not covering the full distance of the first sector. Therefore, the sector and the lap are incomplete and there is no time for them. This is not a limitation of FastF1.

**Comment by psychemedia** [comment]
You can get lap times for all laps, including the first lap, from the ergast API.

For example:

```python
from fastf1.ergast import base_url, _parse_ergast, _parse_json_response, _headers
from fastf1.api import Cache
from pandas import json_normalize
import pandas as pd

def _parse_ergast(data, table="RaceTable:Races", as_df=False):
    """Parse specified table from ergast data."""
    data_table = data['MRData']
    for t in table.split(":"):
        if t:
            data_table = data_table[t]
    return data_table if not as_df else json_normalize(data_table)

#http://ergast.com/api/f1/drivers
#http://ergast.com/api/f1/2010/drivers
#http://ergast.com/api/f1/2010/2/drivers
def fetch_driver(year, gp=None, as_df=False):
    """Get driver information from ergast API by year and optionally by race."""
    if gp is None:
        url = f"{base_url}/{year}/drivers.json"
    else:
        url = f"{base_url}/{year}/{gp}/drivers.json"
    return _parse_ergast(_parse_json_response(
        Cache.requests_get(url, headers=_headers)),
                         "DriverTable:Drivers", as_df
    )

def ergast_driver_code_number(year, gp=None, typ="code", index="driverId"):
    """Get driver codes via ergast API."""
    _driver_details = fetch_driver(year, gp, as_df=True)
    _driver_details["permanentNumber"] = _driver_details["permanentNumber"].astype(int)
    typ = "permanentNumber" if typ.startswith("num") else typ
    if typ in ["code", "permanentNumber"]:
        return _driver_details[["driverId", typ]].set_index(index).squeeze().to_dict()
    return _driver_details[["code", "permanentNumber", "driverId"]]
 

ERGAST_CODES = ergast_driver_code_number(2022)

#http://ergast.com/api/f1/2011/5/laps/1
#http://ergast.com/api/f1/2011/5/drivers/alonso/laps/1
def fetch_laptimes(year, gp, lap=None, driver=None, as_df=False):
    """Get laptime information from ergast API by year and race,
        and optionally by driver and/or lap."""
    url = f"{base_url}/{year}/{gp}"
    if driver:
        url = f"{url}/drivers/{driver}"
    url = f"{url}/laps"
    if lap:
        url = f"{url}/{lap}"
    url = f"{url}.json"
    _laps = _parse_ergast(_parse_json_response(
        Cache.requests_get(url, headers=_headers))
    )[0]["Laps"]
    if as_df:
        _df = pd.DataFrame(_laps[0]["Timings"])
        _df['driverId'] = _df['driverId'].map(ERGAST_CODES)
        return _df
    return _laps

# Get laptimes for first lap for second race of 2022
fetch_laptimes(2022, 2, 1)
```

**Reply by theOehrly** [reply]
You are right, Ergast does provide lap times for the first lap. I stand by my previous explanation above, for now, though. I doubled checked this again, and as far as I can tell, the timing system does not provide a lap time for the first lap. On TV and in the App, the first lap time appears when the drivers finish their second lap.

Now I am curious, what the source for those lap times on Ergast is. I want to know, how accurate the first lap times are and how they are defined. I'd suppose lights out is the zero reference. That wouldn't play well with some of FastF1's current features, because it is assumed that the cars are in the same position (timing line) at the beginning of each lap. But that would be incorrect for the first lap then.

And finally, if we wanted to implement this data, the whole core of FastF1 needs to be made more flexible in terms of merging data from multiple sources. That'd be a larger refactoring operation, although it is necessary anyway.

**Reply by psychemedia** [reply]
I think ergast relies on time sheets  If there's an overall race time report, simple arithmetic will give you lap 1 time, albeit possibly incorporating rounding errors

**Reply by theOehrly** [reply]
> I think ergast relies on time sheets  If there's an overall race time report, simple arithmetic will give you lap 1 time, albeit possibly incorporating rounding errors

Do you know where such time sheets are published?

**Reply by psychemedia** [reply]
Timing sheets  and other docs are available from the FIA: https://www.fia.com/documents/championships/fia-formula-one-world-championship-14/season/season-2022-2005 (hmmm.. there used to be "Event and timing information" section, but I can't see it from a quick glance? Ah: https://www.fia.com/events/fia-formula-one-world-championship/season-2022/formula-one then click on an event, *Show Menu*, then *Event and timing information*; eg https://www.fia.com/events/fia-formula-one-world-championship/season-2022/british-grand-prix/eventtiming-information)

I used to collect timing sheets, but haven't archived them for several years now; my old downloader is [here](https://github.com/f1datajunkie/f1djR/blob/master/scraping/PDF%20scraper%202018.ipynb), I think. There may also be sample PDF scrapers in that repo, but it's been several years since I played with them, and I'm not sure I committed them all at the time.

**Reply by psychemedia** [reply]
I seem to have an archive of FIA docs 2013-2019. I'm not sure if you can find archived docs on the FIA website (It's been a long time since I poked around it...)

**Reply by psychemedia** [reply]
> " I want to know, how accurate the first lap times are and how they are defined."

First lap is really interesting, There are several losses compared to a flying lap, eg:

- time to get up to speed (acceleration from a standing start, cold tyres/brakes, etc);
- reaction time off the line;
- traffic effects, time to get on to a racing line, etc.;
- additional distance travelled by each car relative to front row.

**Reply by psychemedia** [reply]
The FIA lap analysis timing sheet gives the time the car passed the timing line as the first lap time (eg 2022_10_gbr_f1_r0_timing_racelapanalysis_v01.pdf)

<img width="665" alt="image" src="https://user-images.githubusercontent.com/82988/183854455-fe292ed6-981b-4bf4-a06a-46c1e969c48d.png">

So from that you can get eg the time delta at timing line for each car, eg relative to car that crosses timing line first.

The race history chart has all lap times, including lap 1. I'd guess that is the delta from end of lap 1 to race start time. You should be able to confirm that by finding a timing sheet for an event where someone sat on the grid for a couple of seconds..

<img width="680" alt="image" src="https://user-images.githubusercontent.com/82988/183855145-32830e85-8607-468b-9251-cccb3a4ee4f1.png">

**Reply by theOehrly** [reply]
@psychemedia Thank you for summarizing this in such a concise form. I need to take some time and work through this information. And I need to check what we actually can use this data for. Potentially, it could allow to better synchronize laps.
But as I already said, implementing this requires a better strategy for merging data from F1 and Ergast. Furthermore, the order of downloading, post-processing and merging operations might need to be changed to implement this properly. At this point, it becomes a major (but anyway necessary) refactoring operation. I don't just want to hack this into the current FastF1 core. Also, I do anticipate some necessary breaking changes. 
I can see this being part of a new major release, but this very likely won't happen within the current season still.

**Reply by psychemedia** [reply]
It might be better to represent laptimes separate from the provider, so that differemt laptiime providers (ergast, fia pdf scrape, derived from timescreen, etc) can be accommodated

**Reply by theOehrly** [reply]
Can you explain a bit more in detail what you mean by that?

**Reply by psychemedia** [reply]
In functional terms, being able to do something like: `getLaptimes(..., provider='ergast')`

To support introspection of the provider,  class `Laps`, could also have a `self.provider` attribute.

**Reply by theOehrly** [reply]
This has the one major disadvantage, that Ergast as a single provider can by far not provide all the data that the `Laps` class currently makes available. And this is true as well for any other potential provider that I can think of. Apart from empty fields this will also break all functionality related to getting telemetry from laps data. I'm not sure whether this would not create more problems than it solves.

If something like this were to be added, it must be well thought trough, with the appropriate errors and informational messages to guide users. I fully support the idea of unifying different backends/providers through FastF1. But the issue is that these backends partially have very different capabilities. Therefore, there will always be things that stop working when switching between backends/providers.
